src/kpro.erl

```diff
@@ -345,79 +355,100 @@ decode(kpro_FetchResponsePartition, Bin) ->
+          { partition           = Partition
+          , errorCode           = kpro_ErrorCode:decode(ErrorCode)
+          , highWatermarkOffset = HighWmOffset
+          , messageSetSize      = MessageSetSize
+          , message_L           = Messages
+          };
+      true ->
+        #{ partition           => Partition
+         , errorCode           => kpro_ErrorCode:decode(ErrorCode)
+         , highWatermarkOffset => HighWmOffset
+         , messageSetSize      => MessageSetSize
+         , messages            => Messages}
+    end,
   {PartitionMessages, Rest};
-decode(StructName, Bin) when is_atom(StructName) ->
-  kpro_structs:decode(StructName, Bin).
+decode(UseMaps, StructName, Bin) when is_atom(StructName) ->
+  kpro_structs:decode(UseMaps, StructName, Bin).
 
-decode_message_stream(<<>>, Acc) ->
+decode_message_stream(_UseMaps, <<>>, Acc) ->
   Acc;
-decode_message_stream(Bin, Acc) ->
+decode_message_stream(UseMaps, Bin, Acc) ->
   {Msg, Rest} =
-    try decode(kpro_Message, Bin)
+    try kpro_structs:decode(UseMaps, kpro_Message, Bin)
     catch error : {badmatch, _} ->
       {?incomplete_message, <<>>}
     end,
   NewAcc =
-    case Msg of
-      #kpro_Message{attributes = Attr} = Msg when ?KPRO_IS_GZIP_ATTR(Attr) ->
-        decode_message_stream(zlib:gunzip(Msg#kpro_Message.value), Acc);
-      #kpro_Message{attributes = Attr} = Msg when ?KPRO_IS_SNAPPY_ATTR(Attr) ->
-        decode_message_stream(java_snappy_unpack(Msg#kpro_Message.value), Acc);
-      #kpro_Message{attributes = Attr} = Msg when ?KPRO_IS_LZ4_ATTR(Attr) ->
-        decode_message_stream(lz4_unpack(Msg#kpro_Message.value), Acc);
+    case kpro_message:attributes(Msg) of
+      Attr when ?KPRO_IS_GZIP_ATTR(Attr) ->
+        decode_message_stream(UseMaps, zlib:gunzip(kpro_message:value(Msg)), Acc);
+      Attr when ?KPRO_IS_SNAPPY_ATTR(Attr) ->
+        decode_message_stream(UseMaps, java_snappy_unpack(kpro_message:value(Msg)), Acc);
+      Attr when ?KPRO_IS_LZ4_ATTR(Attr) ->
+        decode_message_stream(UseMaps, lz4_unpack(kpro_message:value(Msg)), Acc);
       _Else ->
         [Msg | Acc]
     end,
-  decode_message_stream(Rest, NewAcc).
+  decode_message_stream(UseMaps, Rest, NewAcc).
 
-decode_fields(RecordName, Fields, Bin) ->
-  {FieldValues, BinRest} = do_decode_fields(RecordName, Fields, Bin, _Acc = []),
+decode_fields(false = UseMaps, RecordName, Fields, Bin) ->
+  {FieldValues, BinRest} = do_decode_fields(UseMaps, RecordName, Fields, Bin, _Acc = []),
   %% make the record.
-  {list_to_tuple([RecordName | FieldValues]), BinRest}.
-
-do_decode_fields(_RecordName, _Fields = [], Bin, Acc) ->
+  {list_to_tuple([RecordName | FieldValues]), BinRest};
+decode_fields(true = UseMaps, RecordName, Fields, Bin) ->
+  do_decode_fields(UseMaps, RecordName, Fields, Bin, _Acc = #{}).
+  %{Acc, BinRest} = do_decode_fields(UseMaps, RecordName, Fields, Bin, _Acc = #{}),
+  %[$k, $p, $r, $o, $_ | ShortRecordName] = atom_to_list(RecordName),
```

Comment:
why is it commented out ?
is the a case in which the message tag is useful ?
how about `maps:from_list([{tag, RecordName} | Fields])`